semanticdb3/semanticdb3.md

```diff
@@ -0,0 +1,491 @@
+Language Server Protocol (LSP) [\[2\]][2], a protocol that connects programming
+language implementations and text editors, has gained strong industrial support
+and at the time of writing has implementations for over 25 programming languages.
+Build Server Protocol (BSP) [\[3\]][3] follows in LSP's tracks with an ambition to
+define a protocol for communication between language servers and build tools.
+
+While lots of work in the open-source community has been invested in unifying
+user experience (by codifying commonly used operations like go to definition or
+find all references), relatively less work went into unifying implementor experience.
+For example, at the moment, there exist five different LSP implementations for Scala
+[[4][4], [5][5], [6][6], [7][7], [8][8]]. They all implement the same protocol
+that works with code, but they all use different data structures to represent that code.
+
+Without a standard way to share information between tools, implementors have
+two unpleasant choices. First, they can use compiler internals, which are often
+underdocumented and lack compatibility guarantees. Otherwise, they reimplement
+compiler internals, which usually leads to duplication of effort and inconsistent UX.
+For example, Scala IDE [\[9\]][9] uses Scala compiler internals, which has known
+stability problems in interactive mode. To the contrast, IntelliJ [\[10\]][10] has
+its own Scala typechecker, which is more stable but is known for spurious red squiggles.
+
+This demonstrates the necessity for portable metaprogramming APIs -
+something that we have been working on within Scalameta [\[11\]][11].
+In the previous years, we shipped portable syntactic APIs for Scala,
+including abstract syntax trees, parsing and prettyprinting [\[12\]][12].
+SemanticDB is our take on portable semantic APIs.
+
+## Overview
+
+SemanticDB is a data model for semantic information about programs in Scala and
+other languages. SemanticDB decouples production and consumption of semantic information,
+establishing documented means for communication between tools.
+
+TODO: Once we upgrade semanticdb-scalac to support SemanticDB v3,
+demonstrate a hello world example. Show the original program, a prettyprinted
+version of its SemanticDB data, very briefly explain the sections.
+
+TODO: Once we ship metac and metap that support SemanticDB v3, explain that
+the SemanticDB data model is materialized as Protocol Buffers [\[13\]][13], JSON [\[14\]][14]
+and SQL [\[15\]][15]. Demonstrate how we can use metac to generate `.semanticdb` files
+from the hello world example above, show the binary payload, decode the files with metap.
+
+TODO: Explain how these protobuf files can be used in tools like scalafix, metadoc and metals.
+Mention that beyond protobuf files, there are many ways to use SemanticDB as an
+interchange format, including passing objects around in memory, communication via JSON, etc etc.
+
+## Model
```

Comment:
Data Model

Reply:
I thought about changing `Model` to `Data Model` and `Schemas` to `Data Schemas`, but didn't find that the added clarity is worth the added clutter. What's your take on this?

Reply:
I don't think it adds clutter, jumping straight into the document without context it's not clear what "Model" means.

Reply:
Okay, I'll update.

Reply:
FWIW I like “data model” and schema, but not “data schema”